File: framework/widget/src/main/java/com/sofar/widget/recycler/CenterPagerSnapHelper.java
package com.sofar.widget.recycler;

import android.view.View;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.OrientationHelper;
import androidx.recyclerview.widget.PagerSnapHelper;
import androidx.recyclerview.widget.RecyclerView;

/**
 * 保证子View的中线始终与RecyclerView本身的中线对齐
 * <p>
 * 与PagerSnapHelper的区别在于，不算decoration and margin，保证视觉效果上子View始终在RecyclerView的中间
 */
public class CenterPagerSnapHelper extends PagerSnapHelper {

  @Nullable
  private OrientationHelper mVerticalHelper;
  @Nullable
  private OrientationHelper mHorizontalHelper;

  @Nullable
  @Override
  public int[] calculateDistanceToFinalSnap(@NonNull RecyclerView.LayoutManager layoutManager,
    @NonNull View targetView) {
    int[] out = new int[2];
    if (layoutManager.canScrollHorizontally()) {
      out[0] = distanceToCenter(layoutManager, targetView,
        getHorizontalHelper(layoutManager));
    } else {
      out[0] = 0;
    }

    if (layoutManager.canScrollVertically()) {
      out[1] = distanceToCenter(layoutManager, targetView,
        getVerticalHelper(layoutManager));
    } else {
      out[1] = 0;
    }
    return out;
  }

  private int distanceToCenter(@NonNull RecyclerView.LayoutManager layoutManager,
    @NonNull View targetView, OrientationHelper helper) {
    final int childCenter = targetView.getLeft() + (targetView.getMeasuredWidth() / 2);
    final int containerCenter = helper.getStartAfterPadding() + helper.getTotalSpace() / 2;
    return childCenter - containerCenter;
  }

  @NonNull
  private OrientationHelper getVerticalHelper(@NonNull RecyclerView.LayoutManager layoutManager) {
    if (mVerticalHelper == null) {
      mVerticalHelper = OrientationHelper.createVerticalHelper(layoutManager);
    }
    return mVerticalHelper;
  }

  @NonNull
  private OrientationHelper getHorizontalHelper(
    @NonNull RecyclerView.LayoutManager layoutManager) {
    if (mHorizontalHelper == null) {
      mHorizontalHelper = OrientationHelper.createHorizontalHelper(layoutManager);
    }
    return mHorizontalHelper;
  }
}

File: core/ui/src/main/java/com/sofar/core/ui/recyclerview/CenterPagerSnapHelper.kt
package com.sofar.core.ui.recyclerview

import android.view.View
import androidx.recyclerview.widget.OrientationHelper
import androidx.recyclerview.widget.PagerSnapHelper
import androidx.recyclerview.widget.RecyclerView

/**
 * 保证子View的中线始终与RecyclerView本身的中线对齐
 *
 * 与PagerSnapHelper的区别在于，不算decoration and margin，保证视觉效果上子View始终在RecyclerView的中间
 */
class CenterPagerSnapHelper : PagerSnapHelper() {

  private var mVerticalHelper: OrientationHelper? = null
  private var mHorizontalHelper: OrientationHelper? = null

  override fun calculateDistanceToFinalSnap(
    layoutManager: RecyclerView.LayoutManager,
    targetView: View
  ): IntArray {
    val out = IntArray(2)
    if (layoutManager.canScrollHorizontally()) {
      out[0] = distanceToCenter(
        layoutManager, targetView,
        getHorizontalHelper(layoutManager)
      )
    } else {
      out[0] = 0
    }

    if (layoutManager.canScrollVertically()) {
      out[1] = distanceToCenter(
        layoutManager, targetView,
        getVerticalHelper(layoutManager)
      )
    } else {
      out[1] = 0
    }
    return out
  }

  private fun distanceToCenter(
    layoutManager: RecyclerView.LayoutManager,
    targetView: View,
    helper: OrientationHelper
  ): Int {
    val childCenter = targetView.left + targetView.measuredWidth / 2
    val containerCenter = helper.startAfterPadding + helper.totalSpace / 2
    return childCenter - containerCenter
  }

  private fun getVerticalHelper(layoutManager: RecyclerView.LayoutManager): OrientationHelper {
    return mVerticalHelper ?: OrientationHelper.createVerticalHelper(layoutManager).also {
      mVerticalHelper = it
    }
  }

  private fun getHorizontalHelper(layoutManager: RecyclerView.LayoutManager): OrientationHelper {
    return mHorizontalHelper ?: OrientationHelper.createHorizontalHelper(layoutManager).also {
      mHorizontalHelper = it
    }
  }
}
